src/main/kotlin/org/rust/ide/completion/AttributeCompletionProvider.kt

```diff
@@ -0,0 +1,129 @@
+package org.rust.ide.completion
+
+import com.intellij.codeInsight.completion.CompletionParameters
+import com.intellij.codeInsight.completion.CompletionProvider
+import com.intellij.codeInsight.completion.CompletionResultSet
+import com.intellij.codeInsight.lookup.LookupElementBuilder
+import com.intellij.patterns.ElementPattern
+import com.intellij.patterns.PatternCondition
+import com.intellij.patterns.PlatformPatterns.psiElement
+import com.intellij.patterns.PsiElementPattern
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiFile
+import com.intellij.util.ProcessingContext
+import org.rust.cargo.util.cargoProjectRoot
+import org.rust.lang.RustLanguage
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.util.module
+
+object AttributeCompletionProvider : CompletionProvider<CompletionParameters>() {
+
+    private data class RustAttribute(val name: String, val appliesTo: ElementPattern<PsiElement>)
+
+    val onStruct: PsiElementPattern.Capture<PsiElement> = onItem<RustStructItemElement>()
+
+    val onEnum: PsiElementPattern.Capture<PsiElement> = onItem<RustEnumItemElement>()
+
+    val onFn: PsiElementPattern.Capture<PsiElement> = onItem<RustFnItemElement>()
+
+    val onMod: PsiElementPattern.Capture<PsiElement> = onItem<RustModItemElement>()
+
+    val onStatic: PsiElementPattern.Capture<PsiElement> = onItem<RustStaticItemElement>()
+
+    val onStaticMut: PsiElementPattern.Capture<PsiElement> = psiElement().with(
+        object : PatternCondition<PsiElement>("onStaticMutCondition") {
+            override fun accepts(t: PsiElement, context: ProcessingContext?): Boolean {
+                val elem = t.parent?.parent?.parent
+                return (elem is RustStaticItemElement) && elem.mut != null
+            }
+        })
+
+    val onMacro: PsiElementPattern.Capture<PsiElement> = onItem<RustMacroItemElement>()
+
+    val onTupleStruct: PsiElementPattern.Capture<PsiElement> = psiElement()
+        .withSuperParent(3, psiElement().withChild(psiElement<RustStructTupleArgsElement>()))
+
+    val onCrate: PsiElementPattern.Capture<PsiElement> = psiElement().withSuperParent<PsiFile>(3).with(
+        object : PatternCondition<PsiElement>("onCrateCondition") {
+            override fun accepts(t: PsiElement, context: ProcessingContext?): Boolean {
+                val file = t.containingFile.originalFile.virtualFile
+                val crateSourceRoot = "${t.module?.cargoProjectRoot ?: "temp://"}/src"
+                val crateFile = listOf("$crateSourceRoot/lib.rs", "$crateSourceRoot/main.rs")
+                return crateFile.contains(file.toString())
```

Comment:
Hm, looks like we already have a function to check if the file is a crate root, [here it is](isCrateRoot)! It should be ok to cast `PsiFile` to `RustFile` here. 

[isCrateRoot]: https://github.com/intellij-rust/intellij-rust/blob/master/src/main/kotlin/org/rust/lang/core/psi/impl/RustFile.kt#L43